# What is Adversarial Training?

Adversarial training is a defense strategy in which **adversarial examples are generated during training**, and the model is trained on these perturbed samples.  
By repeatedly exposing the model to attacks, it learns to become **robust** against them.

---

# Why Do We Need It?

Deep neural networks are highly vulnerable to **adversarial perturbations**—small, often imperceptible changes to the input that cause incorrect predictions.

Adversarial training helps the model:

- Learn a smoother and more stable loss landscape  
- Become more resistant to adversarial attacks  
- Improve robustness (usually at a slight cost to clean accuracy)

---

# PGD Adversarial Training (Madry et al.)

The most widely used and strongest baseline for adversarial robustness is **PGD adversarial training**. In this way, for each batch during training:

1. Start with a slightly perturbed version of the input  
2. Apply multiple PGD steps to craft a strong adversarial example  
3. Train the model using these adversarial samples  

This training update is commonly expressed as:
$
\theta_{t+1} = \theta_t - \eta \, \nabla_{\theta} \, L\big(f_{\theta}(x_{\text{adv}}),\, y\big)
$, where $x_{\text{adv}}$ was created by PGD.


# Adversarial Training Scenarios

Below are the different training strategies explored in this notebook. Each scenario demonstrates a distinct way of incorporating adversarial examples into the training process.

Let's begin by importing the necessary libraries and setting up initial configurations:




In [3]:
import copy
import math
import random
import time
from tqdm import tqdm
from typing import Tuple, Callable

import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
from torch.utils.data import DataLoader, Subset

import torchvision
import torchvision.transforms as transforms
from torchvision.models import resnet18

In [4]:
# --- Reproducibility helpers ---
def set_seed(seed: int = 0):
    random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    torch.backends.cudnn.deterministic = True
    torch.backends.cudnn.benchmark = False

# --- Device ---
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

# --- Seed Setting (!!! do not change the seed !!!) ---
SEED = 2025
set_seed(SEED)

# Data / training hyperparameters
BATCH_SIZE = 128
LR = 0.1
MOMENTUM = 0.9
WEIGHT_DECAY = 5e-4

# 20 clean epochs, 20 std-adv epochs
EPOCHS_CLEAN = 20
EPOCHS_ADV_STD = 20      # standard PGD AT (for counting budget)
EPOCHS_ADV_MAX = 50      # max epochs for other ADV variants
EPOCHS_SEQ_CLEAN = 20    # clean phase epochs for sequential

EPS = 8 / 255.0
PGD_STEPS = 8
PGD_STEP_SIZE = 2 / 255.0
M = 4

In [5]:
# --- Data preparation ---
transform_train = transforms.Compose([
                                      transforms.RandomCrop(32, padding=4),
                                      transforms.RandomHorizontalFlip(),
                                      transforms.ToTensor()]
                                     )
transform_test = transforms.Compose([transforms.ToTensor()])

trainset = torchvision.datasets.CIFAR10(root='./data', train=True, download=True, transform=transform_train)
testset = torchvision.datasets.CIFAR10(root='./data', train=False, download=True, transform=transform_test)

indices_train = torch.randperm(len(trainset), generator=torch.Generator().manual_seed(SEED))[:30000]
indices_test  = torch.randperm(len(testset), generator=torch.Generator().manual_seed(SEED))[:5000]

train_subset = Subset(trainset, indices_train)
test_subset = Subset(testset, indices_test)

trainloader = DataLoader(train_subset, batch_size=BATCH_SIZE, shuffle=True, num_workers=2)
testloader = DataLoader(test_subset, batch_size=BATCH_SIZE, shuffle=False, num_workers=2)

100%|██████████| 170M/170M [00:03<00:00, 49.3MB/s] 


In [7]:
# --- Model & Optimizer helpers ---
def make_model(num_classes=10):
    # TODO
    model = resnet18(pretrained=False)
    model.conv1 = nn.Conv2d(3, 64, kernel_size=3, stride=1, padding=1, bias=False)
    model.maxpool = nn.Identity()
    model.fc = nn.Linear(model.fc.in_features, num_classes)

    return model

def make_optimizer(model):
    # TODO
    optimizer = optim.SGD(model.parameters(), lr=LR, momentum=MOMENTUM, weight_decay=WEIGHT_DECAY)

    return optimizer

**Hint:** In our adversarial training setup, the goal of the budget system is to measure or constrain **only the expensive computations**, namely the gradient-based operations performed during PGD steps and the model’s backward pass. Since `.backward()` and `autograd.grad()` calls dominate the computational cost of adversarial training, the budget should be tied specifically to those operations.

In **standard adversarial training**, where we simply want to track how much compute is used without interrupting training, we use **budget_mode="count"** and call `budget_add` every time a gradient is computed.

In contrast, when we want to enforce a strict compute limit we use **budget_mode="consume"**, calling `budget_sub` after each gradient computation and stopping early once the budget is exhausted.



In [8]:
# --- Budget helpers ---
def make_budget(initial=0):
    """Create a mutable budget object."""
    return {"value": int(initial)}

def budget_add(budget, n=1):
    """Increases the budget counter."""
    if budget is not None:
        budget["value"] += int(n)

def budget_sub(budget, n=1):
    """Decreases the budget counter."""
    if budget is not None:
        budget["value"] -= int(n)

def budget_left(budget):
    """Returns the remaining budget, or None when budget is disabled."""
    return None if budget is None else int(budget["value"])

def budget_exhausted(budget):
    """Checks whether the budget is finished."""
    return (budget is not None) and (budget["value"] <= 0)

In [9]:
# --- PGD attack implementation ---
def pgd_attack(
    model: nn.Module,
    x: torch.Tensor,
    y: torch.Tensor,
    eps: float = EPS,
    alpha: float = PGD_STEP_SIZE,
    steps: int = PGD_STEPS,
    budget=None,
    mode: str = None,
) -> torch.Tensor:
    """
    model : model to attack
    x : clean input batch
    y : labels for x
    eps : max L∞ perturbation
    alpha : PGD step size
    steps : number of PGD iterations
    budget : optional budget object for AT
    mode:
      - None      : ignore budget (for evaluation, etc.)
      - "count"   : budget++ for every autograd.grad (std adv training)
      - "consume" : budget-- for every autograd.grad, stop early if budget <= 0
    """
    x_adv = ...

    # TODO: Implement PGD attack with budget calculation
    x_adv = x.detach().clone()
    x_adv += torch.empty_like(x_adv).uniform_(-eps, eps)
    x_adv = torch.clamp(x_adv, 0, 1).detach()

    for step in range(steps):
        # Check budget if in consume mode
        if mode == "consume" and budget_exhausted(budget):
            break

        x_adv.requires_grad = True

        # Forward pass
        logits = model(x_adv)
        loss = F.cross_entropy(logits, y)

        # Compute gradient with respect to input
        grad = torch.autograd.grad(loss, x_adv, retain_graph=False, create_graph=False)[0]

        # Update budget based on mode
        if mode == "count":
            budget_add(budget, 1)
        elif mode == "consume":
            budget_sub(budget, 1)

        # PGD step: gradient ascent on input
        x_adv = x_adv.detach() + alpha * grad.sign()

        # Project back to eps ball around original input
        delta = torch.clamp(x_adv - x, -eps, eps)
        x_adv = torch.clamp(x + delta, 0, 1).detach()

    return x_adv

In [10]:
# --- Evaluation functions ---
def evaluate_clean(model: nn.Module, dataloader: DataLoader) -> Tuple[float, float]:
    model.eval()
    correct = 0
    total = 0

    # TODO: Evaluate the model on clean samples
    with torch.no_grad():
        for x, y in dataloader:
            x, y = x.to(device), y.to(device)
            logits = model(x)
            pred = logits.argmax(dim=1)
            correct += (pred == y).sum().item()
            total += y.size(0)

    acc = correct / total
    return acc

def evaluate_adv(model: nn.Module, dataloader: DataLoader, attack_fn: Callable = pgd_attack, eps: float = EPS) -> Tuple[float, float]:
    model.eval()
    correct = 0
    total = 0

    # TODO: Apply PGD attack on clean data and evaluate the model on perturbed samples
    for x, y in dataloader:
        x, y = x.to(device), y.to(device)

        # Generate adversarial examples (no budget tracking in evaluation)
        x_adv = attack_fn(model, x, y, eps=eps, mode=None)

        with torch.no_grad():
            logits = model(x_adv)
            pred = logits.argmax(dim=1)
            correct += (pred == y).sum().item()
            total += y.size(0)

    acc = correct / total
    return acc

In [11]:
# ---Prepare identical initialization ---
set_seed(SEED)
base_model = make_model().to(device)
initial_state = copy.deepcopy(base_model.state_dict())

/usr/local/lib/python3.12/dist-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=None`.
  warnings.warn(msg)


## 1) Baseline: Train a Clean Model, Then Apply PGD Attack

In this scenario, we train a model **only on clean data**, with no adversarial examples involved.  
After training, we evaluate the model under a PGD attack.

In this scenario, our goal is to show how a standard (non-robust) model suffers *accuracy collapse* when exposed to strong adversarial perturbations. This serves as the reference point against which all adversarial training methods are compared.

In [12]:
def train_one_epoch_clean(model, loader, optimizer, budget=None):
    """
    Standard clean training on clean data only.
    """
    total_loss = 0.0

    # TODO: Implement one epoch clean training with budget calculation
    model.train()

    for x, y in loader:
        x, y = x.to(device), y.to(device)

        optimizer.zero_grad()
        logits = model(x)
        loss = F.cross_entropy(logits, y)
        loss.backward()

        # Count backward pass in budget if provided
        if budget is not None:
            budget_add(budget, 1)

        optimizer.step()
        total_loss += loss.item()

    total_loss = total_loss / len(loader)

    return total_loss

In [13]:
# Train clean model
model_clean = make_model().to(device)
model_clean.load_state_dict(copy.deepcopy(initial_state))
opt_clean = make_optimizer(model_clean)

print('Training clean baseline model...')
for ep in tqdm(range(1, EPOCHS_CLEAN + 1)):
    loss = train_one_epoch_clean(model_clean, trainloader, opt_clean)
    print(f"Epoch {ep}/{EPOCHS_CLEAN} | clean loss: {loss:.4f}")

torch.save(model_clean.state_dict(), "model_clean.pth")

clean_acc = evaluate_clean(model_clean, testloader)
print(f"\nClean model test accuracy: {clean_acc:.4f}")

Training clean baseline model...


  5%|▌         | 1/20 [00:17<05:37, 17.78s/it]

Epoch 1/20 | clean loss: 2.1135


 10%|█         | 2/20 [00:34<05:09, 17.20s/it]

Epoch 2/20 | clean loss: 1.5996


 15%|█▌        | 3/20 [00:51<04:49, 17.02s/it]

Epoch 3/20 | clean loss: 1.3770


 20%|██        | 4/20 [01:08<04:30, 16.92s/it]

Epoch 4/20 | clean loss: 1.1702


 25%|██▌       | 5/20 [01:24<04:13, 16.88s/it]

Epoch 5/20 | clean loss: 0.9978


 30%|███       | 6/20 [01:41<03:55, 16.84s/it]

Epoch 6/20 | clean loss: 0.8617


 35%|███▌      | 7/20 [01:58<03:38, 16.82s/it]

Epoch 7/20 | clean loss: 0.7529


 40%|████      | 8/20 [02:15<03:21, 16.81s/it]

Epoch 8/20 | clean loss: 0.6768


 45%|████▌     | 9/20 [02:32<03:04, 16.81s/it]

Epoch 9/20 | clean loss: 0.6163


 50%|█████     | 10/20 [02:48<02:48, 16.81s/it]

Epoch 10/20 | clean loss: 0.5714


 55%|█████▌    | 11/20 [03:05<02:31, 16.81s/it]

Epoch 11/20 | clean loss: 0.5496


 60%|██████    | 12/20 [03:22<02:14, 16.80s/it]

Epoch 12/20 | clean loss: 0.5255


 65%|██████▌   | 13/20 [03:39<01:57, 16.81s/it]

Epoch 13/20 | clean loss: 0.5058


 70%|███████   | 14/20 [03:56<01:40, 16.80s/it]

Epoch 14/20 | clean loss: 0.4836


 75%|███████▌  | 15/20 [04:12<01:24, 16.80s/it]

Epoch 15/20 | clean loss: 0.4683


 80%|████████  | 16/20 [04:29<01:07, 16.80s/it]

Epoch 16/20 | clean loss: 0.4579


 85%|████████▌ | 17/20 [04:46<00:50, 16.81s/it]

Epoch 17/20 | clean loss: 0.4442


 90%|█████████ | 18/20 [05:03<00:33, 16.81s/it]

Epoch 18/20 | clean loss: 0.4353


 95%|█████████▌| 19/20 [05:20<00:16, 16.81s/it]

Epoch 19/20 | clean loss: 0.4248


100%|██████████| 20/20 [05:37<00:00, 16.85s/it]

Epoch 20/20 | clean loss: 0.4200



Clean model test accuracy: 0.7388


In [14]:
adv_acc = evaluate_adv(model_clean, testloader, eps=8/255.0)
print(f"Under PGD attack (8/255) -> adversarial accuracy: {adv_acc:.4f}")

Under PGD attack (8/255) -> adversarial accuracy: 0.0000


### Now let's apply a weaker PGD attack to this cleanly trained model and observe the difference…

In [15]:
adv_acc_eps4 = evaluate_adv(model_clean, testloader, eps=4/255.0)
print(f"Under PGD attack (4/255) -> adversarial accuracy: {adv_acc_eps4:.4f}")

Under PGD attack (4/255) -> adversarial accuracy: 0.0012


In [16]:
adv_acc_eps2 = evaluate_adv(model_clean, testloader, eps=2/255.0)
print(f"Under PGD attack (2/255) -> adversarial accuracy: {adv_acc_eps2:.4f}")

Under PGD attack (2/255) -> adversarial accuracy: 0.0486


### **Question:** What can you infer from the adversarial accuracy presented above?
**Your Answer:** The clean model shows severe vulnerability to adversarial attacks, with accuracy dropping dramatically even under small perturbations, demonstrating the need for adversarial training.

Based on the results (Clean accuracy: 0.7388, PGD-8/255: 0.0000, PGD-4/255: 0.0012, PGD-2/255: 0.0486), we can infer:

1. **Complete vulnerability under strong attacks**: The model achieves 73.88% accuracy on clean data but drops to **0% accuracy** under the standard 8/255 PGD attack. This represents a **catastrophic failure** where the model cannot correctly classify even a single adversarial example. This demonstrates that deep neural networks, despite high clean performance, are extremely fragile to adversarial perturbations.

2. **Attack strength is critical**: Even when we reduce the perturbation budget:
   - At 4/255 (half the perturbation): adversarial accuracy is only 0.12% (still essentially zero)
   - At 2/255 (quarter perturbation): adversarial accuracy improves to 4.86%, but this is still a **93.4% drop** from clean accuracy
   
   This shows that even very small, imperceptible perturbations (2/255 ≈ 0.8% of the pixel range) can fool the model over 93% of the time.

3. **Fundamental lack of robustness**: Standard neural networks learn decision boundaries that are highly non-smooth and rely on brittle, non-robust features. They exploit spurious statistical correlations in the training data rather than learning semantically meaningful representations. This makes them trivially exploitable by gradient-based adversarial attacks.

4. **The robustness problem is severe**: The near-zero adversarial accuracy reveals that adversarial vulnerability is not just a minor issue but a fundamental problem that requires dedicated defense mechanisms like adversarial training to address.

This baseline clearly establishes the critical need for adversarial training techniques to build models that can maintain reasonable performance under adversarial attacks.



## 2) Standard Adversarial Training (PGD)

Using the **same initial weights as the clean model**, we train the model using **PGD-generated adversarial examples** at every step.

**How it works:**
1. For each batch, run PGD to generate $ x_{\text{adv}} $.  
2. Compute the loss on these adversarial samples.  
3. Backpropagate and update model parameters.

In this scenarion, we want to obtain the classical *PGD-adversarially trained model*.


In [17]:
def train_one_epoch_adv_standard(
    model,
    loader,
    optimizer,
    attack_fn=pgd_attack,
    budget=None,
    budget_mode=None,   # "count" or "consume" or None
):
    """
    Standard PGD adversarial training on adversarial examples only.
    model : neural network to train
    loader : training dataloader
    optimizer : optimizer used for updating model weights
    attack_fn : adversarial attack function (default: pgd_attack)
    budget : optional budget object for tracking compute
    budget_mode :
    - When budget_mode=="count": budget++ for autograd.grad in PGD and for backward().
    - When budget_mode=="consume": budget-- for both and stop when exhausted.
    """
    total_loss = 0.0

    # TODO: Implement Standard PGD adversarial training with budget calculation
    model.train()

    for x, y in loader:
        # Check if budget exhausted (for consume mode)
        if budget_mode == "consume" and budget_exhausted(budget):
            break

        x, y = x.to(device), y.to(device)

        # Generate adversarial examples
        x_adv = attack_fn(model, x, y, budget=budget, mode=budget_mode)

        # Train on adversarial examples
        optimizer.zero_grad()
        logits = model(x_adv)
        loss = F.cross_entropy(logits, y)
        loss.backward()

        # Count or consume budget for backward pass
        if budget_mode == "count":
            budget_add(budget, 1)
        elif budget_mode == "consume":
            budget_sub(budget, 1)

        optimizer.step()
        total_loss += loss.item()

    total_loss = total_loss / len(loader)

    return total_loss

In [18]:
# --- Standard adversarial training starting from SAME initial weights ---
model_adv_standard = make_model().to(device)
model_adv_standard.load_state_dict(copy.deepcopy(initial_state))
opt_adv_std = make_optimizer(model_adv_standard)
std_budget = make_budget(0) # This budget will COUNT all backward() and autograd.grad() calls

print('Training standard adversarial training (PGD) from same init...')
for ep in tqdm(range(1, EPOCHS_ADV_STD + 1)):
    loss = train_one_epoch_adv_standard(model_adv_standard, trainloader, opt_adv_std,
                                        attack_fn=pgd_attack, budget=std_budget, budget_mode="count")
    print(f"Epoch {ep}/{EPOCHS_ADV_STD} | adv loss: {loss:.4f}")

total_budget = budget_left(std_budget)
print(f"\nTotal gradient budget from standard PGD AT: {total_budget}")

torch.save(model_adv_standard.state_dict(), "model_adv_standard.pth")

Training standard adversarial training (PGD) from same init...


  5%|▌         | 1/20 [01:51<35:10, 111.07s/it]

Epoch 1/20 | adv loss: 2.3532


 10%|█         | 2/20 [03:42<33:18, 111.05s/it]

Epoch 2/20 | adv loss: 2.0877


 15%|█▌        | 3/20 [05:33<31:28, 111.06s/it]

Epoch 3/20 | adv loss: 2.0159


 20%|██        | 4/20 [07:24<29:36, 111.06s/it]

Epoch 4/20 | adv loss: 1.9665


 25%|██▌       | 5/20 [09:15<27:45, 111.05s/it]

Epoch 5/20 | adv loss: 1.9252


 30%|███       | 6/20 [11:06<25:54, 111.06s/it]

Epoch 6/20 | adv loss: 1.8886


 35%|███▌      | 7/20 [12:57<24:03, 111.07s/it]

Epoch 7/20 | adv loss: 1.8597


 40%|████      | 8/20 [14:48<22:12, 111.08s/it]

Epoch 8/20 | adv loss: 1.8190


 45%|████▌     | 9/20 [16:39<20:21, 111.08s/it]

Epoch 9/20 | adv loss: 1.7802


 50%|█████     | 10/20 [18:30<18:30, 111.08s/it]

Epoch 10/20 | adv loss: 1.7494


 55%|█████▌    | 11/20 [20:21<16:39, 111.08s/it]

Epoch 11/20 | adv loss: 1.7201


 60%|██████    | 12/20 [22:12<14:48, 111.06s/it]

Epoch 12/20 | adv loss: 1.6861


 65%|██████▌   | 13/20 [24:03<12:57, 111.06s/it]

Epoch 13/20 | adv loss: 1.6690


 70%|███████   | 14/20 [25:54<11:06, 111.07s/it]

Epoch 14/20 | adv loss: 1.6401


 75%|███████▌  | 15/20 [27:45<09:15, 111.05s/it]

Epoch 15/20 | adv loss: 1.6260


 80%|████████  | 16/20 [29:37<07:24, 111.05s/it]

Epoch 16/20 | adv loss: 1.6069


 85%|████████▌ | 17/20 [31:28<05:33, 111.04s/it]

Epoch 17/20 | adv loss: 1.5943


 90%|█████████ | 18/20 [33:19<03:42, 111.04s/it]

Epoch 18/20 | adv loss: 1.5786


 95%|█████████▌| 19/20 [35:10<01:51, 111.03s/it]

Epoch 19/20 | adv loss: 1.5653


100%|██████████| 20/20 [37:01<00:00, 111.06s/it]

Epoch 20/20 | adv loss: 1.5514

Total gradient budget from standard PGD AT: 42300


In [20]:
acc_clean_std = evaluate_clean(model_adv_standard, testloader)
acc_adv_std = evaluate_adv(model_adv_standard, testloader, eps=8/255.0)
print(f"\nStandard adv-trained model -> clean acc: {acc_clean_std:.4f} | adv acc: {acc_adv_std:.4f}")


Standard adv-trained model -> clean acc: 0.6512 | adv acc: 0.3886


### Let's examine how the attack’s epsilon ball influences the robust model’s accuracy...


In [21]:
acc_adv_std_eps4 = evaluate_adv(model_adv_standard, testloader, eps=4/255.0)
print(f"Under PGD attack (4/255) -> adversarial accuracy: {acc_adv_std_eps4:.4f}")

Under PGD attack (4/255) -> adversarial accuracy: 0.5232


In [22]:
acc_adv_std_eps2 = evaluate_adv(model_adv_standard, testloader, eps=2/255.0)
print(f"Under PGD attack (2/255) -> adversarial accuracy: {acc_adv_std_eps2:.4f}")

Under PGD attack (2/255) -> adversarial accuracy: 0.5888


## 3) Joint Clean + Adversarial Loss

For each clean batch:
1. Generate adversarial examples using PGD or FGSM.  
2. Compute **clean loss** and **adversarial loss**.  
3. Backpropagate on their **mean**:
$
L_{\text{total}} = \frac{1}{2}\big(L(x, y) + L(x_{\text{adv}}, y)\big)
$

In this case we want to train a model that balances *clean accuracy* and *robust accuracy*, by learning from both types of data simultaneously.


In [23]:
def train_one_epoch_mean_clean_adv(
    model,
    loader,
    optimizer,
    attack_fn=pgd_attack,
    budget=None,
):
    """
    One backward per batch on 0.5*(loss_clean + loss_adv).
    In adv training (mean), we *consume* budget: both PGD grads and backward() are charged.
    """
    total_loss = 0.0

    # TODO: Implement PGD adversarial training based on the Mean Loss with budget calculation
    model.train()

    for x, y in loader:
        # Check if budget exhausted
        if budget_exhausted(budget):
            break

        x, y = x.to(device), y.to(device)

        # Generate adversarial examples (consumes budget)
        x_adv = attack_fn(model, x, y, budget=budget, mode="consume")

        if budget_exhausted(budget):
            break

        # Compute clean loss
        optimizer.zero_grad()
        logits_clean = model(x)
        loss_clean = F.cross_entropy(logits_clean, y)

        # Compute adversarial loss
        logits_adv = model(x_adv)
        loss_adv = F.cross_entropy(logits_adv, y)

        # Mean loss
        loss = 0.5 * (loss_clean + loss_adv)
        loss.backward()

        # Consume budget for backward
        budget_sub(budget, 1)

        optimizer.step()
        total_loss += loss.item()

    total_loss = total_loss / len(loader)

    return total_loss

In [24]:
# --- Mean (clean, adv) under fixed budget ---
model_mean = make_model().to(device)
model_mean.load_state_dict(copy.deepcopy(initial_state))
opt_mean = make_optimizer(model_mean)
mean_budget = make_budget(total_budget)

print('Training on mean(clean loss, adv loss) under fixed budget...')
for ep in tqdm(range(1, EPOCHS_ADV_MAX + 1)):
    if budget_exhausted(mean_budget):
        print(f"Budget exhausted before epoch {ep}.")
        break

    loss = train_one_epoch_mean_clean_adv(model_mean, trainloader, opt_mean,
                                          attack_fn=pgd_attack, budget=mean_budget)
    print(
        f"Epoch {ep}/{EPOCHS_ADV_MAX} | mean loss: {loss:.4f} "
        f"| budget left: {budget_left(mean_budget)}"
    )

    if budget_exhausted(mean_budget):
        print("Budget exhausted, stopping mean training.")
        break

torch.save(model_mean.state_dict(), "model_mean.pth")

Training on mean(clean loss, adv loss) under fixed budget...


  2%|▏         | 1/50 [02:06<1:43:40, 126.94s/it]

Epoch 1/50 | mean loss: 2.3115 | budget left: 40185


  4%|▍         | 2/50 [04:13<1:41:32, 126.93s/it]

Epoch 2/50 | mean loss: 1.9544 | budget left: 38070


  6%|▌         | 3/50 [06:20<1:39:26, 126.94s/it]

Epoch 3/50 | mean loss: 1.8643 | budget left: 35955


  8%|▊         | 4/50 [08:27<1:37:19, 126.95s/it]

Epoch 4/50 | mean loss: 1.7808 | budget left: 33840


 10%|█         | 5/50 [10:34<1:35:12, 126.94s/it]

Epoch 5/50 | mean loss: 1.7000 | budget left: 31725


 12%|█▏        | 6/50 [12:41<1:33:06, 126.96s/it]

Epoch 6/50 | mean loss: 1.6302 | budget left: 29610


 14%|█▍        | 7/50 [14:48<1:30:59, 126.97s/it]

Epoch 7/50 | mean loss: 1.5724 | budget left: 27495


 16%|█▌        | 8/50 [16:55<1:28:52, 126.96s/it]

Epoch 8/50 | mean loss: 1.5152 | budget left: 25380


 18%|█▊        | 9/50 [19:02<1:26:45, 126.96s/it]

Epoch 9/50 | mean loss: 1.4687 | budget left: 23265


 20%|██        | 10/50 [21:09<1:24:38, 126.96s/it]

Epoch 10/50 | mean loss: 1.4223 | budget left: 21150


 22%|██▏       | 11/50 [23:16<1:22:31, 126.95s/it]

Epoch 11/50 | mean loss: 1.3807 | budget left: 19035


 24%|██▍       | 12/50 [25:23<1:20:24, 126.95s/it]

Epoch 12/50 | mean loss: 1.3471 | budget left: 16920


 26%|██▌       | 13/50 [27:30<1:18:17, 126.97s/it]

Epoch 13/50 | mean loss: 1.3197 | budget left: 14805


 28%|██▊       | 14/50 [29:37<1:16:11, 126.97s/it]

Epoch 14/50 | mean loss: 1.3002 | budget left: 12690


 30%|███       | 15/50 [31:44<1:14:04, 126.98s/it]

Epoch 15/50 | mean loss: 1.2754 | budget left: 10575


 32%|███▏      | 16/50 [33:51<1:11:57, 126.99s/it]

Epoch 16/50 | mean loss: 1.2553 | budget left: 8460


 34%|███▍      | 17/50 [35:58<1:09:50, 127.00s/it]

Epoch 17/50 | mean loss: 1.2424 | budget left: 6345


 36%|███▌      | 18/50 [38:05<1:07:44, 127.00s/it]

Epoch 18/50 | mean loss: 1.2252 | budget left: 4230


 38%|███▊      | 19/50 [40:12<1:05:36, 126.99s/it]

Epoch 19/50 | mean loss: 1.2115 | budget left: 2115


 38%|███▊      | 19/50 [42:19<1:09:03, 133.65s/it]

Epoch 20/50 | mean loss: 1.2030 | budget left: 0
Budget exhausted, stopping mean training.


In [25]:
acc_clean_mean = evaluate_clean(model_mean, testloader)
acc_adv_mean = evaluate_adv(model_mean, testloader)
print(f"\nMean-loss model -> clean acc: {acc_clean_mean:.4f} | adv acc: {acc_adv_mean:.4f}")


Mean-loss model -> clean acc: 0.7188 | adv acc: 0.3476


## 4) Sequential Training: Clean First → Adversarial Later

Training is split into two distinct phases:

- **Phase 1:** Train the model normally on clean data only.  
- **Phase 2:** Continue training using only adversarial examples.

Here we start with a well-performing clean model, then adapt it to robustness.  


In [26]:
# --- Sequential training: clean phase then adversarial phase ---
model_seq = make_model().to(device)
model_seq.load_state_dict(copy.deepcopy(initial_state))
opt_seq = make_optimizer(model_seq)
seq_budget = make_budget(total_budget)

print('Sequential training (clean -> adv) under fixed budget...')

# 1) Clean phase: fixed number of epochs, consume budget per backward
for ep_clean in tqdm(range(1, EPOCHS_SEQ_CLEAN + 1)):
    if budget_exhausted(seq_budget):
        print(f"Budget exhausted during clean phase before epoch {ep_clean}.")
        break

    # loss_clean = train_one_epoch_clean(model_seq, trainloader, opt_seq) # we did not use this because it doesn't control budget!
    model_seq.train()
    loss_clean = 0.0

    for x, y in trainloader:
        if budget_exhausted(seq_budget):
            break

        x, y = x.to(device), y.to(device)

        opt_seq.zero_grad()
        logits = model_seq(x)
        loss = F.cross_entropy(logits, y)
        loss.backward()

        # Consume budget for backward
        budget_sub(seq_budget, 1)

        opt_seq.step()
        loss_clean += loss.item()

    loss_clean /= len(trainloader)

    print(
        f"Sequential clean phase | epoch {ep_clean}/{EPOCHS_SEQ_CLEAN} "
        f"| loss: {loss_clean:.4f} | budget left: {budget_left(seq_budget)}"
    )

    if budget_exhausted(seq_budget):
        print("Budget exhausted at end of clean phase.")
        break

print('Switching to adversarial phase...')

# 2) Adversarial phase: PGD training consuming whatever budget is left
for ep_adv in tqdm(range(1, EPOCHS_ADV_MAX + 1)):
    if budget_exhausted(seq_budget):
        print(f"Budget exhausted before adversarial epoch {ep_adv}.")
        break

    loss_adv = train_one_epoch_adv_standard(model_seq, trainloader, opt_seq,
                                            attack_fn=pgd_attack, budget=seq_budget,
                                            budget_mode="consume")

    print(
        f"Sequential adv phase | epoch {ep_adv}/{EPOCHS_ADV_MAX} "
        f"| loss: {loss_adv:.4f} | budget left: {budget_left(seq_budget)}"
    )

    if budget_exhausted(seq_budget):
        print("Budget exhausted in adversarial phase, stopping.")
        break

torch.save(model_seq.state_dict(), "model_seq.pth")


Sequential training (clean -> adv) under fixed budget...


  5%|▌         | 1/20 [00:16<05:19, 16.79s/it]

Sequential clean phase | epoch 1/20 | loss: 1.9645 | budget left: 42065


 10%|█         | 2/20 [00:33<05:02, 16.78s/it]

Sequential clean phase | epoch 2/20 | loss: 1.5037 | budget left: 41830


 15%|█▌        | 3/20 [00:50<04:45, 16.80s/it]

Sequential clean phase | epoch 3/20 | loss: 1.2486 | budget left: 41595


 20%|██        | 4/20 [01:07<04:28, 16.80s/it]

Sequential clean phase | epoch 4/20 | loss: 1.0443 | budget left: 41360


 25%|██▌       | 5/20 [01:23<04:11, 16.79s/it]

Sequential clean phase | epoch 5/20 | loss: 0.9284 | budget left: 41125


 30%|███       | 6/20 [01:40<03:55, 16.80s/it]

Sequential clean phase | epoch 6/20 | loss: 0.8211 | budget left: 40890


 35%|███▌      | 7/20 [01:57<03:38, 16.80s/it]

Sequential clean phase | epoch 7/20 | loss: 0.7331 | budget left: 40655


 40%|████      | 8/20 [02:14<03:21, 16.80s/it]

Sequential clean phase | epoch 8/20 | loss: 0.6523 | budget left: 40420


 45%|████▌     | 9/20 [02:31<03:04, 16.81s/it]

Sequential clean phase | epoch 9/20 | loss: 0.6062 | budget left: 40185


 50%|█████     | 10/20 [02:47<02:48, 16.80s/it]

Sequential clean phase | epoch 10/20 | loss: 0.5712 | budget left: 39950


 55%|█████▌    | 11/20 [03:04<02:31, 16.81s/it]

Sequential clean phase | epoch 11/20 | loss: 0.5346 | budget left: 39715


 60%|██████    | 12/20 [03:21<02:14, 16.81s/it]

Sequential clean phase | epoch 12/20 | loss: 0.5164 | budget left: 39480


 65%|██████▌   | 13/20 [03:38<01:57, 16.81s/it]

Sequential clean phase | epoch 13/20 | loss: 0.4877 | budget left: 39245


 70%|███████   | 14/20 [03:55<01:40, 16.83s/it]

Sequential clean phase | epoch 14/20 | loss: 0.4785 | budget left: 39010


 75%|███████▌  | 15/20 [04:12<01:24, 16.83s/it]

Sequential clean phase | epoch 15/20 | loss: 0.4681 | budget left: 38775


 80%|████████  | 16/20 [04:29<01:07, 16.84s/it]

Sequential clean phase | epoch 16/20 | loss: 0.4468 | budget left: 38540


 85%|████████▌ | 17/20 [04:45<00:50, 16.85s/it]

Sequential clean phase | epoch 17/20 | loss: 0.4462 | budget left: 38305


 90%|█████████ | 18/20 [05:02<00:33, 16.86s/it]

Sequential clean phase | epoch 18/20 | loss: 0.4313 | budget left: 38070


 95%|█████████▌| 19/20 [05:19<00:16, 16.85s/it]

Sequential clean phase | epoch 19/20 | loss: 0.4187 | budget left: 37835


100%|██████████| 20/20 [05:36<00:00, 16.82s/it]


Sequential clean phase | epoch 20/20 | loss: 0.4127 | budget left: 37600
Switching to adversarial phase...


  2%|▏         | 1/50 [01:51<1:30:41, 111.06s/it]

Sequential adv phase | epoch 1/50 | loss: 2.2190 | budget left: 35485


  4%|▍         | 2/50 [03:42<1:28:49, 111.03s/it]

Sequential adv phase | epoch 2/50 | loss: 1.8508 | budget left: 33370


  6%|▌         | 3/50 [05:33<1:26:56, 111.00s/it]

Sequential adv phase | epoch 3/50 | loss: 1.7294 | budget left: 31255


  8%|▊         | 4/50 [07:23<1:25:05, 110.99s/it]

Sequential adv phase | epoch 4/50 | loss: 1.6747 | budget left: 29140


 10%|█         | 5/50 [09:14<1:23:13, 110.97s/it]

Sequential adv phase | epoch 5/50 | loss: 1.6357 | budget left: 27025


 12%|█▏        | 6/50 [11:05<1:21:22, 110.97s/it]

Sequential adv phase | epoch 6/50 | loss: 1.6036 | budget left: 24910


 14%|█▍        | 7/50 [12:56<1:19:31, 110.96s/it]

Sequential adv phase | epoch 7/50 | loss: 1.5823 | budget left: 22795


 16%|█▌        | 8/50 [14:47<1:17:40, 110.96s/it]

Sequential adv phase | epoch 8/50 | loss: 1.5637 | budget left: 20680


 18%|█▊        | 9/50 [16:38<1:15:49, 110.96s/it]

Sequential adv phase | epoch 9/50 | loss: 1.5461 | budget left: 18565


 20%|██        | 10/50 [18:29<1:13:58, 110.95s/it]

Sequential adv phase | epoch 10/50 | loss: 1.5282 | budget left: 16450


 22%|██▏       | 11/50 [20:20<1:12:07, 110.95s/it]

Sequential adv phase | epoch 11/50 | loss: 1.5255 | budget left: 14335


 24%|██▍       | 12/50 [22:11<1:10:16, 110.95s/it]

Sequential adv phase | epoch 12/50 | loss: 1.5103 | budget left: 12220


 26%|██▌       | 13/50 [24:02<1:08:25, 110.95s/it]

Sequential adv phase | epoch 13/50 | loss: 1.5057 | budget left: 10105


 28%|██▊       | 14/50 [25:53<1:06:34, 110.95s/it]

Sequential adv phase | epoch 14/50 | loss: 1.4886 | budget left: 7990


 30%|███       | 15/50 [27:44<1:04:43, 110.94s/it]

Sequential adv phase | epoch 15/50 | loss: 1.4816 | budget left: 5875


 32%|███▏      | 16/50 [29:35<1:02:51, 110.94s/it]

Sequential adv phase | epoch 16/50 | loss: 1.4721 | budget left: 3760


 34%|███▍      | 17/50 [31:26<1:01:01, 110.94s/it]

Sequential adv phase | epoch 17/50 | loss: 1.4662 | budget left: 1645


 34%|███▍      | 17/50 [32:52<1:03:49, 116.05s/it]

Sequential adv phase | epoch 18/50 | loss: 1.1361 | budget left: -1
Budget exhausted in adversarial phase, stopping.


In [27]:
acc_clean_seq = evaluate_clean(model_seq, testloader)
acc_adv_seq = evaluate_adv(model_seq, testloader)
print(f"\nSequential model -> clean acc: {acc_clean_seq:.4f} | adv acc: {acc_adv_seq:.4f}")


Sequential model -> clean acc: 0.7062 | adv acc: 0.4026


## 5) Alternating Training: Clean Batch ↔ Adversarial Batch

During each epoch, training alternates between:

- one clean batch  
- one adversarial batch  
- one clean batch  
- one adversarial batch  
- …and so on

In this scenario, we expose the model to both clean and adversarial data *within the same training phase*, allowing it to maintain clean performance while becoming robust.

In [28]:
def train_one_epoch_alternating(model, loader, optimizer, budget=None):
    """
    For each batch:
      - one clean step
      - one adv step (with PGD)
    Both steps consume budget (PGD autograd.grad + both backwards).
    """
    total_loss_clean = 0.0
    total_loss_adv = 0.0

    # TODO: Implement PGD alternating training with budget calculation
    model.train()

    for x, y in loader:
        if budget_exhausted(budget):
            break

        x, y = x.to(device), y.to(device)

        # --- Clean step ---
        optimizer.zero_grad()
        logits_clean = model(x)
        loss_clean = F.cross_entropy(logits_clean, y)
        loss_clean.backward()

        # Consume budget for clean backward
        budget_sub(budget, 1)

        optimizer.step()
        total_loss_clean += loss_clean.item()

        if budget_exhausted(budget):
            break

        # --- Adversarial step ---
        x_adv = pgd_attack(model, x, y, budget=budget, mode="consume")

        if budget_exhausted(budget):
            break

        optimizer.zero_grad()
        logits_adv = model(x_adv)
        loss_adv = F.cross_entropy(logits_adv, y)
        loss_adv.backward()

        # Consume budget for adv backward
        budget_sub(budget, 1)

        optimizer.step()
        total_loss_adv += loss_adv.item()

    total_loss_clean = total_loss_clean / len(loader)
    total_loss_adv = total_loss_adv / len(loader)

    return (total_loss_clean, total_loss_adv)

In [29]:
# --- Alternating training: one clean batch, one adv batch ---
model_alt = make_model().to(device)
model_alt.load_state_dict(copy.deepcopy(initial_state))
opt_alt = make_optimizer(model_alt)
alt_budget = make_budget(total_budget)

print('Alternating training (clean + adv batches) under fixed budget...')
for ep in tqdm(range(1, EPOCHS_ADV_MAX + 1)):
    if budget_exhausted(alt_budget):
        print(f"Budget exhausted before epoch {ep}.")
        break

    clean_loss, adv_loss = train_one_epoch_alternating(model_alt, trainloader, opt_alt,
                                                        budget=alt_budget)

    print(
        f"Epoch {ep}/{EPOCHS_ADV_MAX} | clean loss: {clean_loss:.4f} "
        f"| adv loss: {adv_loss:.4f} | budget left: {budget_left(alt_budget)}"
    )

    if budget_exhausted(alt_budget):
        print("Budget exhausted, stopping alternating training.")
        break

torch.save(model_alt.state_dict(), "model_alt.pth")


Alternating training (clean + adv batches) under fixed budget...


  2%|▏         | 1/50 [02:07<1:43:46, 127.08s/it]

Epoch 1/50 | clean loss: 2.0538 | adv loss: 2.2782 | budget left: 39950


  4%|▍         | 2/50 [04:14<1:41:41, 127.12s/it]

Epoch 2/50 | clean loss: 1.6794 | adv loss: 2.0546 | budget left: 37600


  6%|▌         | 3/50 [06:21<1:39:34, 127.12s/it]

Epoch 3/50 | clean loss: 1.5348 | adv loss: 1.9893 | budget left: 35250


  8%|▊         | 4/50 [08:28<1:37:28, 127.14s/it]

Epoch 4/50 | clean loss: 1.4145 | adv loss: 1.9398 | budget left: 32900


 10%|█         | 5/50 [10:35<1:35:22, 127.16s/it]

Epoch 5/50 | clean loss: 1.3131 | adv loss: 1.8885 | budget left: 30550


 12%|█▏        | 6/50 [12:42<1:33:15, 127.18s/it]

Epoch 6/50 | clean loss: 1.2246 | adv loss: 1.8504 | budget left: 28200


 14%|█▍        | 7/50 [14:50<1:31:10, 127.21s/it]

Epoch 7/50 | clean loss: 1.1571 | adv loss: 1.8166 | budget left: 25850


 16%|█▌        | 8/50 [16:57<1:29:03, 127.23s/it]

Epoch 8/50 | clean loss: 1.1029 | adv loss: 1.7863 | budget left: 23500


 18%|█▊        | 9/50 [19:04<1:26:57, 127.25s/it]

Epoch 9/50 | clean loss: 1.0637 | adv loss: 1.7663 | budget left: 21150


 20%|██        | 10/50 [21:12<1:24:50, 127.26s/it]

Epoch 10/50 | clean loss: 1.0300 | adv loss: 1.7462 | budget left: 18800


 22%|██▏       | 11/50 [23:19<1:22:42, 127.25s/it]

Epoch 11/50 | clean loss: 1.0027 | adv loss: 1.7289 | budget left: 16450


 24%|██▍       | 12/50 [25:26<1:20:35, 127.26s/it]

Epoch 12/50 | clean loss: 0.9696 | adv loss: 1.7100 | budget left: 14100


 26%|██▌       | 13/50 [27:33<1:18:28, 127.27s/it]

Epoch 13/50 | clean loss: 0.9577 | adv loss: 1.7015 | budget left: 11750


 28%|██▊       | 14/50 [29:41<1:16:21, 127.26s/it]

Epoch 14/50 | clean loss: 0.9378 | adv loss: 1.6887 | budget left: 9400


 30%|███       | 15/50 [31:48<1:14:14, 127.26s/it]

Epoch 15/50 | clean loss: 0.9284 | adv loss: 1.6796 | budget left: 7050


 32%|███▏      | 16/50 [33:55<1:12:07, 127.27s/it]

Epoch 16/50 | clean loss: 0.9086 | adv loss: 1.6684 | budget left: 4700


 34%|███▍      | 17/50 [36:02<1:09:59, 127.26s/it]

Epoch 17/50 | clean loss: 0.8945 | adv loss: 1.6620 | budget left: 2350


 34%|███▍      | 17/50 [38:10<1:14:05, 134.71s/it]

Epoch 18/50 | clean loss: 0.8858 | adv loss: 1.6480 | budget left: 0
Budget exhausted, stopping alternating training.


In [30]:
acc_clean_alt = evaluate_clean(model_alt, testloader)
acc_adv_alt = evaluate_adv(model_alt, testloader)
print(f"\nAlternating model -> clean acc: {acc_clean_alt:.4f} | adv acc: {acc_adv_alt:.4f}")


Alternating model -> clean acc: 0.7358 | adv acc: 0.3346


In [31]:
# --- Summary (print final table) ---
from tabulate import tabulate

rows = [
    ['Baseline clean (no adv training)', f"{clean_acc:.4f}", f"{adv_acc:.4f}"],
    ['Standard PGD adv-train', f"{acc_clean_std:.4f}", f"{acc_adv_std:.4f}"],
    ['Mean(clean, adv)', f"{acc_clean_mean:.4f}", f"{acc_adv_mean:.4f}"],
    ['Sequential (clean -> adv)', f"{acc_clean_seq:.4f}", f"{acc_adv_seq:.4f}"],
    ['Alternating batches', f"{acc_clean_alt:.4f}", f"{acc_adv_alt:.4f}"],
]
print('\nFinal comparison table:')
print(tabulate(rows, headers=['Scenario', 'Clean Acc', 'Adv Acc']))


Final comparison table:
Scenario                            Clean Acc    Adv Acc
--------------------------------  -----------  ---------
Baseline clean (no adv training)       0.7388     0
Standard PGD adv-train                 0.6512     0.3886
Mean(clean, adv)                       0.7188     0.3476
Sequential (clean -> adv)              0.7062     0.4026
Alternating batches                    0.7358     0.3346


### **Question:** Based on the results from all above adversarial training scenarios, compare the key features of each method — including clean accuracy, adversarial accuracy (under PGD), robust–clean accuracy trade-off, and training stability. Which adversarial training strategy provides the best balance between clean and robust performance?

**Your Answer:** As we see in the above table the **'Sequential (clean -> adv)'** strategy provides the best balance between clean and robust performance with *'Clean Acc = 70.62%'* and  *'Adv Acc = 40.26%'*.


Based on the experimental results, here is a detailed comparison:

**1. Standard PGD Adversarial Training (Clean: 0.6512, Adv: 0.3886):**
- Achieves the **second-highest adversarial accuracy** (38.86%)
- Suffers from the **largest clean accuracy drop**: 73.88% → 65.12% (11.9% decrease)
- Most computationally expensive (serves as our budget baseline)
- Training is stable and straightforward with consistent convergence
- Strong robustness but at a considerable cost to clean performance
- Represents the traditional approach that prioritizes robustness over clean accuracy

**2. Mean (Clean + Adversarial) Loss (Clean: 0.7188, Adv: 0.3476):**
- **Second-highest clean accuracy** among adversarial training methods (71.88%)
- Achieves moderate adversarial accuracy (34.76%)
- Moderate clean accuracy retention: 2.7% drop from baseline (73.88% → 71.88%)
- Adversarial accuracy 10.6% lower than standard PGD adversarial training (34.76% vs 38.86%)
- Provides a reasonable balance through joint optimization of both objectives
- Training is stable since both objectives are optimized together
- Middle-ground performance on both metrics

**3. Sequential Training (Clean → Adversarial) (Clean: 0.7062, Adv: 0.4026):**
- **Achieves the highest adversarial accuracy** (40.26%) - the most robust model!
- Good clean accuracy (70.62%), with 4.4% drop from baseline
- Outperforms standard PGD adversarial training in robustness (40.26% vs 38.86%) while using the same budget
- The two-phase approach allows the model to first learn good representations, then adapt them for robustness
- **Best robustness-to-clean-accuracy ratio**: Achieves strongest robustness with only moderate clean accuracy sacrifice
- Training stability can vary at the phase transition point but overall very effective

**4. Alternating Training (Clean: 0.7358, Adv: 0.3346):**
- **Highest clean accuracy** among all adversarial training methods (73.58%), nearly matching baseline (only 0.4% drop)
- However, achieves the **lowest adversarial accuracy** (33.46%) among all adversarial training methods
- Excellent clean performance preservation but weakest robustness
- Clean accuracy drop of only 0.4% is impressive, but adversarial accuracy is 16.7% lower than sequential training
- The alternating approach preserves clean performance but fails to build strong robustness
- Training can be **unstable** due to switching between different objectives each batch
- **Poor efficiency**: Despite excellent clean accuracy, the computational budget is not effectively converted into robustness

**Key Insights:**

1. **Robustness-Accuracy Trade-off**: All adversarial training methods reduce clean accuracy compared to the baseline, but the magnitude varies significantly (0.4% to 11.9% drop). However, smaller clean accuracy drops don't guarantee better robustness—Alternating training has the smallest clean drop but weakest robustness.

2. **Sequential Training Dominates in Robustness**: Sequential training achieving the highest adversarial accuracy (40.26%) while maintaining strong clean accuracy (70.62%, only 4.4% drop) is remarkable. This 4.4% sacrifice yields the strongest robustness, making it the most **efficient** use of the clean-accuracy budget.

3. **Alternating Training's Paradox**: Despite achieving the highest clean accuracy (73.58%, only 0.4% drop from baseline), alternating training has the worst robustness (33.46%). This reveals a critical insight: preserving clean accuracy doesn't automatically translate to robustness. The frequent objective switching prevents the model from learning truly robust features, resulting in a model that performs well on clean data but remains vulnerable to attacks.

4. **Standard PGD adversarial training's Heavy Cost**: Traditional PGD adversarial training achieves good robustness (38.86%) but requires the largest clean accuracy sacrifice (11.9% drop). This makes it less efficient than sequential training, which achieves even better robustness (40.26%) with much smaller clean accuracy loss (4.4% drop).

5. **Efficiency Ranking** (Robustness gained per % of clean accuracy lost):
   - **Sequential**: 40.26% robustness for 4.4% clean loss = **9.15 ratio** (most efficient)
   - **Alternating**: 33.46% robustness for 0.4% clean loss = 83.65 ratio (but absolute robustness is lowest)
   - **Mean Loss**: 34.76% robustness for 2.7% clean loss = **12.87 ratio** (good efficiency)
   - **Standard PGD adversarial training**: 38.86% robustness for 11.9% clean loss = **3.27 ratio** (least efficient)

**Conclusion:**

The **best strategy depends on the application priorities**:

- **For maximum robustness** (recommended for most scenarios): **Sequential Training** (Clean→Adv) is the clear winner, achieving **40.26% adversarial accuracy** (highest) while maintaining **70.62% clean accuracy** (only 4.4% drop). This represents the **best robustness-to-clean-accuracy trade-off** and is ideal for security-critical applications where strong defense against adversarial attacks is essential.

- **For preserving clean accuracy above all**: **Alternating Training** achieves **73.58% clean accuracy** (nearly baseline performance, only 0.4% drop) but sacrifices robustness significantly (33.46%, lowest among adversarial methods). Only suitable when adversarial attacks are rare, mild, or when clean performance is absolutely critical and some robustness is acceptable but not prioritized.

- **For moderate balance with stable training**: **Mean (Clean + Adversarial) Loss** provides **71.88% clean accuracy** with **34.76% adversarial accuracy**. Good efficiency (12.87 ratio) with stable, straightforward training. Suitable for applications where both metrics matter roughly equally and training stability is important.

- **Standard PGD Training**: Achieves 65.12% clean accuracy and 38.86% adversarial accuracy—good robustness but the largest clean accuracy sacrifice (11.9% drop) and lowest efficiency (3.27 ratio). Less recommended given that sequential training achieves better robustness with smaller clean accuracy cost.

**Overall Recommendation**: **Sequential Training** provides the best overall performance, achieving the **strongest robustness (40.26%)** with only a **moderate clean accuracy sacrifice (4.4% drop to 70.62%)**. It successfully balances both objectives more efficiently than any other method, making it the **preferred choice for most practical adversarial training scenarios** where robustness is important. If clean accuracy must be preserved at nearly any cost, alternating training is an option, but be aware it provides significantly weaker defense against adversarial attacks.





# Adversarial Training For Free ([Shafahi et al.](https://arxiv.org/abs/1904.12843))

**Free Adversarial Training** is an efficient method designed to achieve adversarial robustness **without increasing training cost** compared to standard (clean) training.

Traditional adversarial training (especially PGD-based) is expensive because it requires multiple gradient steps **per batch** to generate adversarial examples.  
Free Adversarial Training avoids this overhead using two key ideas:

#### **1) Reuse the same batch multiple times**
Instead of performing many PGD steps *inside a single batch*, the same batch is processed for **multiple mini-steps** (called “free” steps), effectively simulating PGD iterations across repeated passes.

#### **2) Use gradient ascent on the input simultaneously with gradient descent on the weights**
Each training step does two things:

- **Update the adversarial example**  
  by performing gradient ascent on the input (using the backward pass you already computed)

- **Update the model parameters**  
  using the same backward pass (gradient descent as usual)

This means **one backward pass per step**, instead of many per PGD attack. Let's implement this method:


In [32]:
def train_one_epoch_free(
    model,
    loader,
    optimizer,
    m: int = 4,
    eps: float = EPS,
    alpha: float = PGD_STEP_SIZE,
    budget=None,
):
    total_loss = 0.0

    # TODO: Implement PGD adversarial training for free with budget calculation
    # Free adversarial training: reuse the same batch m times, 
    #   updating both the adversarial perturbation and model weights
    #   in each mini-step using a single backward pass.
    model.train()
    
    for x, y in loader:
        x, y = x.to(device), y.to(device)
        
        # Initialize perturbation (randomly within eps ball)
        delta = torch.zeros_like(x).uniform_(-eps, eps)
        delta.requires_grad = True
        
        # Process same batch m times
        for _ in range(m):
            if budget_exhausted(budget):
                break
            
            # Create adversarial example
            x_adv = torch.clamp(x + delta, 0, 1)
            
            # Forward pass
            optimizer.zero_grad()
            logits = model(x_adv)
            loss = F.cross_entropy(logits, y)
            
            # Single backward pass - computes gradients for both model and delta
            loss.backward()
            
            # Consume budget for backward
            budget_sub(budget, 1)
            
            # Update model parameters (gradient descent)
            optimizer.step()
            
            # Update perturbation (gradient ascent on input)
            if delta.grad is not None:
                with torch.no_grad():
                    # Gradient ascent step
                    delta.data = delta.data + alpha * delta.grad.sign()
                    # Project back to eps ball
                    delta.data = torch.clamp(delta.data, -eps, eps)
                    # Zero out gradients for next iteration
                    delta.grad.zero_()
            
            total_loss += loss.item()
        
        if budget_exhausted(budget):
            break
    
    # Average over total mini-steps (batches * m)
    total_loss = total_loss / (len(loader) * m)

    return total_loss

In [33]:
# --- Free adversarial training under fixed budget ---
model_adv_free = make_model().to(device)
model_adv_free.load_state_dict(copy.deepcopy(initial_state))
opt_adv_free = make_optimizer(model_adv_free)
free_budget = make_budget(total_budget)

print(f'Training Free-AT with m={M} from same init under fixed budget...')
for ep in tqdm(range(1, EPOCHS_ADV_MAX + 1)):
    if budget_exhausted(free_budget):
        print(f"Budget exhausted before epoch {ep}.")
        break

    loss = train_one_epoch_free(model_adv_free, trainloader, opt_adv_free, 
                                m=M, eps=EPS, alpha=PGD_STEP_SIZE, budget=free_budget)
    print(
        f"Epoch {ep}/{EPOCHS_ADV_MAX} | free-adv loss: {loss:.4f} "
        f"| budget left: {budget_left(free_budget)}"
    )

    if budget_exhausted(free_budget):
        print("Budget exhausted, stopping Free-AT training.")
        break

torch.save(model_adv_free.state_dict(), "model_adv_free.pth")

Training Free-AT with m=4 from same init under fixed budget...


  2%|▏         | 1/50 [01:06<54:12, 66.39s/it]

Epoch 1/50 | free-adv loss: 2.0399 | budget left: 41360


  4%|▍         | 2/50 [02:12<53:05, 66.37s/it]

Epoch 2/50 | free-adv loss: 1.7326 | budget left: 40420


  6%|▌         | 3/50 [03:19<51:58, 66.35s/it]

Epoch 3/50 | free-adv loss: 1.6064 | budget left: 39480


  8%|▊         | 4/50 [04:25<50:51, 66.34s/it]

Epoch 4/50 | free-adv loss: 1.5072 | budget left: 38540


 10%|█         | 5/50 [05:31<49:46, 66.36s/it]

Epoch 5/50 | free-adv loss: 1.4253 | budget left: 37600


 12%|█▏        | 6/50 [06:38<48:40, 66.38s/it]

Epoch 6/50 | free-adv loss: 1.3626 | budget left: 36660


 14%|█▍        | 7/50 [07:44<47:34, 66.38s/it]

Epoch 7/50 | free-adv loss: 1.3211 | budget left: 35720


 16%|█▌        | 8/50 [08:50<46:27, 66.36s/it]

Epoch 8/50 | free-adv loss: 1.2966 | budget left: 34780


 18%|█▊        | 9/50 [09:57<45:20, 66.36s/it]

Epoch 9/50 | free-adv loss: 1.2538 | budget left: 33840


 20%|██        | 10/50 [11:03<44:13, 66.35s/it]

Epoch 10/50 | free-adv loss: 1.2351 | budget left: 32900


 22%|██▏       | 11/50 [12:09<43:07, 66.34s/it]

Epoch 11/50 | free-adv loss: 1.2153 | budget left: 31960


 24%|██▍       | 12/50 [13:16<42:00, 66.33s/it]

Epoch 12/50 | free-adv loss: 1.1966 | budget left: 31020


 26%|██▌       | 13/50 [14:22<40:54, 66.33s/it]

Epoch 13/50 | free-adv loss: 1.1926 | budget left: 30080


 28%|██▊       | 14/50 [15:28<39:47, 66.33s/it]

Epoch 14/50 | free-adv loss: 1.1856 | budget left: 29140


 30%|███       | 15/50 [16:35<38:41, 66.33s/it]

Epoch 15/50 | free-adv loss: 1.1863 | budget left: 28200


 32%|███▏      | 16/50 [17:41<37:35, 66.34s/it]

Epoch 16/50 | free-adv loss: 1.1739 | budget left: 27260


 34%|███▍      | 17/50 [18:47<36:29, 66.34s/it]

Epoch 17/50 | free-adv loss: 1.1598 | budget left: 26320


 36%|███▌      | 18/50 [19:54<35:22, 66.34s/it]

Epoch 18/50 | free-adv loss: 1.1557 | budget left: 25380


 38%|███▊      | 19/50 [21:00<34:16, 66.34s/it]

Epoch 19/50 | free-adv loss: 1.1504 | budget left: 24440


 40%|████      | 20/50 [22:06<33:10, 66.34s/it]

Epoch 20/50 | free-adv loss: 1.1384 | budget left: 23500


 42%|████▏     | 21/50 [23:13<32:03, 66.34s/it]

Epoch 21/50 | free-adv loss: 1.1370 | budget left: 22560


 44%|████▍     | 22/50 [24:19<30:57, 66.35s/it]

Epoch 22/50 | free-adv loss: 1.1425 | budget left: 21620


 46%|████▌     | 23/50 [25:25<29:51, 66.34s/it]

Epoch 23/50 | free-adv loss: 1.1377 | budget left: 20680


 48%|████▊     | 24/50 [26:32<28:44, 66.34s/it]

Epoch 24/50 | free-adv loss: 1.1356 | budget left: 19740


 50%|█████     | 25/50 [27:38<27:38, 66.33s/it]

Epoch 25/50 | free-adv loss: 1.1389 | budget left: 18800


 52%|█████▏    | 26/50 [28:44<26:32, 66.34s/it]

Epoch 26/50 | free-adv loss: 1.1328 | budget left: 17860


 54%|█████▍    | 27/50 [29:51<25:25, 66.34s/it]

Epoch 27/50 | free-adv loss: 1.1330 | budget left: 16920


 56%|█████▌    | 28/50 [30:57<24:19, 66.34s/it]

Epoch 28/50 | free-adv loss: 1.1227 | budget left: 15980


 58%|█████▊    | 29/50 [32:03<23:13, 66.34s/it]

Epoch 29/50 | free-adv loss: 1.1196 | budget left: 15040


 60%|██████    | 30/50 [33:10<22:06, 66.34s/it]

Epoch 30/50 | free-adv loss: 1.1279 | budget left: 14100


 62%|██████▏   | 31/50 [34:16<21:01, 66.37s/it]

Epoch 31/50 | free-adv loss: 1.1202 | budget left: 13160


 64%|██████▍   | 32/50 [35:23<19:54, 66.38s/it]

Epoch 32/50 | free-adv loss: 1.1227 | budget left: 12220


 66%|██████▌   | 33/50 [36:29<18:48, 66.38s/it]

Epoch 33/50 | free-adv loss: 1.1181 | budget left: 11280


 68%|██████▊   | 34/50 [37:35<17:42, 66.39s/it]

Epoch 34/50 | free-adv loss: 1.1173 | budget left: 10340


 70%|███████   | 35/50 [38:42<16:35, 66.39s/it]

Epoch 35/50 | free-adv loss: 1.1225 | budget left: 9400


 72%|███████▏  | 36/50 [39:48<15:29, 66.40s/it]

Epoch 36/50 | free-adv loss: 1.1152 | budget left: 8460


 74%|███████▍  | 37/50 [40:55<14:23, 66.40s/it]

Epoch 37/50 | free-adv loss: 1.1185 | budget left: 7520


 76%|███████▌  | 38/50 [42:01<13:16, 66.41s/it]

Epoch 38/50 | free-adv loss: 1.1112 | budget left: 6580


 78%|███████▊  | 39/50 [43:07<12:10, 66.40s/it]

Epoch 39/50 | free-adv loss: 1.1132 | budget left: 5640


 80%|████████  | 40/50 [44:14<11:03, 66.40s/it]

Epoch 40/50 | free-adv loss: 1.1233 | budget left: 4700


 82%|████████▏ | 41/50 [45:20<09:57, 66.39s/it]

Epoch 41/50 | free-adv loss: 1.1120 | budget left: 3760


 84%|████████▍ | 42/50 [46:27<08:51, 66.39s/it]

Epoch 42/50 | free-adv loss: 1.1061 | budget left: 2820


 86%|████████▌ | 43/50 [47:33<07:44, 66.40s/it]

Epoch 43/50 | free-adv loss: 1.1112 | budget left: 1880


 88%|████████▊ | 44/50 [48:39<06:38, 66.39s/it]

Epoch 44/50 | free-adv loss: 1.1062 | budget left: 940


 88%|████████▊ | 44/50 [49:46<06:47, 67.87s/it]

Epoch 45/50 | free-adv loss: 1.1057 | budget left: 0
Budget exhausted, stopping Free-AT training.


In [34]:
acc_clean_free = evaluate_clean(model_adv_free, testloader)
acc_adv_free = evaluate_adv(model_adv_free, testloader)
print(f"\nAdversarial Training For Free model -> clean acc: {acc_clean_free:.4f} | adv acc: {acc_adv_free:.4f}")


Adversarial Training For Free model -> clean acc: 0.5532 | adv acc: 0.1296
